src/grep.rs

```diff
@@ -197,25 +203,28 @@ fn grep_offset(
 /// Run bgrep with the given options, outputting to the given `StdoutLock`.
 /// Error detail may be outputted to stderr.
 /// Returns whether there was a match.
-fn run_file(
-  stdout: &mut io::StdoutLock,
+fn run_file<O: Write, P: Into<PathBuf>, B: AsMut<Vec<u8>>>(
+  stdout: &mut O,
   options: &args::Options,
   pattern: &Regex,
-  path: PathBuf,
-  buffer: &mut Vec<u8>,
+  path: P,
+  buffer: &mut B
 ) -> io::Result<bool> {
+  let buffer = buffer.as_mut();
+  let path = path.into();
+
   buffer.clear();
 
   let (read_result, path) =
     if path == Path::new(args::STDIN) { // Path::new is cost-free.
       (io::stdin().lock().read_to_end(buffer), Path::new("<stdin>").display())
     }
     else {
-      let mut file = File::open(&path)
-                          .map_err(|e| {
-                            eprintln!("Error: failed to open file '{}'", path.display());
-                            e
-                          })?;
+      let file = &mut File::open(&path)
```

Comment:
I believe a reference has no benefit here. If I'm right, could you revert this please?